File: python/easy/coffee-machine/coffee_machine.py
import enum


class CoffeeChoice(enum.Enum):
    ESPRESSO = 1
    LATTE = 2
    CAPPUCCINO = 3


class Coffee:
    def __init__(self, money=0, water=0, milk=0, beans=0, cups=1):
        self.money = money
        self.water = water
        self.milk = milk
        self.beans = beans
        self.cups = cups

    def __sub__(self, other):
        money = self.money + other.money
        water = self.water - other.water
        milk = self.milk - other.milk
        beans = self.beans - other.beans
        cups = self.cups - other.cups
        return Coffee(money, water, milk, beans, cups)


class CoffeeMaker:
    def __init__(self):
        self.milk = 540
        self.water = 400
        self.beans = 120

        self.info = Coffee(550, 400, 540, 120, 9)
        self.espresso = Coffee(money=4, water=250, beans=16)
        self.latte = Coffee(money=7, water=350, milk=75, beans=20)
        self.cappuccino = Coffee(money=6, water=200, milk=100, beans=12)

    def __make(self):
        milk_for_coffee = self.milk // self.milk
        water_for_coffee = self.water // self.water
        beans_for_coffee = self.beans // self.coffee_beans
        return min(milk_for_coffee, water_for_coffee, beans_for_coffee)

    def remaining(self):
        print("The coffee machine has:")
        print(f"{self.info.water} of water")
        print(f"{self.info.milk} of milk")
        print(f"{self.info.beans} of coffee beans")
        print(f"{self.info.cups} of disposable cups")
        print(f"${self.info.money} of money" if self.info.money > 0 else f"{self.info.money} of money")
        print()

    def menu(self):
        while True:
            choice = input("Write action (buy, fill, take, remaining, exit)")
            print()
            if choice == "exit":
                break
            if choice == "buy":
                self.buy()
            if choice == "fill":
                self.fill()
            if choice == "take":
                self.take()
            if choice == "remaining":
                self.remaining()

    def had_enough(self, coffee):
        if self.info.milk < coffee.milk:
            return "Sorry, not enough milk!"
        if self.info.water < coffee.water:
            return "Sorry, not enough water!"
        if self.info.beans < coffee.beans:
            return "Sorry, not enough coffee beans!"
        if self.info.cups < coffee.cups:
            return "Sorry, not enough disposable cups!"
        self.info -= coffee
        return "I have enough resources, making you a coffee!"


    def buy(self):
        coffee_choice = input("What do you want to buy? 1 - espresso, 2 - latte, 3 - cappuccino:")
        if coffee_choice == "back":
            return
        coffee_choice = CoffeeChoice(int(coffee_choice))
        if coffee_choice == CoffeeChoice.ESPRESSO:
            coffee = self.espresso
        if coffee_choice == CoffeeChoice.LATTE:
            coffee = self.latte
        if coffee_choice == CoffeeChoice.CAPPUCCINO:
            coffee = self.cappuccino
        print(self.had_enough(coffee))


    def fill(self):
        self.info.water +=  int(input("Write how many ml of water do you want to add:"))
        self.info.milk +=   int(input("Write how many ml of milk do you want to add:"))
        self.info.beans +=  int(input("Write how many grams of coffee beans do you want to add:"))
        self.info.cups +=   int(input("Write how many disposable cups of coffee do you want to add:"))

    def take(self):
        print(f"I gave you {self.info.money}")
        self.info.money = 0


if __name__ == '__main__':
    coffee_maker = CoffeeMaker()
    coffee_maker.menu()

File: java/easy/coffee-machine/CoffeeMachine.java
package machine;
import java.util.Scanner;

class Coffee {
    public int money;
    public int water;
    public int milk;
    public int beans;
    public int cups;

    public Coffee(int money, int water, int milk, int beans, int cups) {
        this.money = money;
        this.water = water;
        this.milk = milk;
        this.beans = beans;
        this.cups = cups;
    }

    public Coffee subtract(Coffee other) {
        int money = this.money + other.money;
        int water = this.water - other.water;
        int milk = this.milk - other.milk;
        int beans = this.beans - other.beans;
        int cups = this.cups - other.cups;
        return new Coffee(money, water, milk, beans, cups);
    }
}

public class CoffeeMachine {
    private Coffee info;
    private final Scanner scanner;
    public int water;
    public int milk;
    public int beans;

    public CoffeeMachine() {
        this.info = new Coffee(550, 400, 540, 120, 9);
        this.water = 400;
        this.milk = 540;
        this.beans = 120;
        this.scanner = new Scanner(System.in);
    }

    private void remaining() {
        System.out.print("The coffee machine has:\n");
        System.out.printf("%d of water\n", this.info.water);
        System.out.printf("%d of milk\n", this.info.milk);
        System.out.printf("%d of coffee beans\n", this.info.beans);
        System.out.printf("%d of disposable cups\n", this.info.cups);
        System.out.printf("$%d of money\n", this.info.money);
    }

    private void menu() {
        while (true) {
            System.out.println("Write action (buy, fill, take, remaining, exit):");
            String choice = scanner.nextLine();
            System.out.println();
            if (choice.equals("buy"))
                this.buy();
            if (choice.equals("fill"))
                this.fill();
            if (choice.equals("take"))
                this.take();
            if (choice.equals("remaining"))
                this.remaining();
            if (choice.equals("exit"))
                break;
        }
    }

    private void take() {
        System.out.printf("I gave you $%d\n", this.info.money);
        this.info.money = 0;
    }

    private void fill() {
        System.out.println("Write how many ml of water do you want to add:");
        this.info.water += this.scanner.nextInt();
        System.out.println("Write how many ml of milk do you want to add:");
        this.info.milk += this.scanner.nextInt();
        System.out.println("Write how many grams of coffee beans do you want to add:");
        this.info.beans += this.scanner.nextInt();
        System.out.println("Write how many disposable cups of coffee do you want to add:");
        this.info.cups += this.scanner.nextInt();
    }

    private void buy() {
        System.out.println("What do you want to buy? 1 - espresso, 2 - latte, 3 - cappuccino, back - to main menu:");
        String coffeeChoice = this.scanner.nextLine();
        Coffee coffee = new Coffee(0, 0, 0, 0, 0);
        if (coffeeChoice.equals("back")) return;
        if (coffeeChoice.equals("1")) coffee = new Coffee(4, 250, 0, 16, 1);
        if (coffeeChoice.equals("2")) coffee = new Coffee(7, 350, 75, 20, 1);
        if (coffeeChoice.equals("3")) coffee = new Coffee(6, 200, 100, 12, 1);
        System.out.println(this.hadEnough(coffee));
    }

    private String hadEnough(Coffee coffee) {
        if (this.info.milk < coffee.milk)
            return "Sorry, not enough milk!";
        if (this.info.water < coffee.water)
            return "Sorry, not enough water!";
        if (this.info.beans < coffee.beans)
            return "Sorry, not enough coffee beans!";
        if (this.info.cups < coffee.cups)
            return "Sorry, not enough disposable cups!";
        this.info = this.info.subtract(coffee);
        return "I have enough resources, making you a coffee!";
    }

    public static void main(String[] args) {
        CoffeeMachine machine = new CoffeeMachine();
        machine.menu();
    }
}
